icu4c/source/i18n/calendar.cpp

```diff
@@ -1561,19 +1561,53 @@ void Calendar::computeFields(UErrorCode &ec)
     // Compute time-related fields.  These are independent of the date and
     // of the subclass algorithm.  They depend only on the local zone
     // wall milliseconds in day.
-    int32_t millisInDay =  (int32_t) (localMillis - (days * kOneDay));
+
+    // For doubles x and n, where n is an integer and ⌊x+n⌋ < 2³¹, the
+    // expression `(int32_t) (x + n)` evaluated with rounding to nearest
+    // differs from ⌊x+n⌋ if 0 < ⌈x⌉−x ≪ x+n, as `x + n` is rounded up to
+    // n+⌈x⌉ = ⌊x+n⌋ + 1.  Rewriting it as ⌊x⌋+n makes the addition exact.
+    int32_t millisInDay =  (int32_t) (uprv_floor(localMillis) - (days * kOneDay));
```

Comment:
Suggestion: To get `days`, up above, there is

```cpp
    int32_t days =  (int32_t)ClockMath::floorDivide(localMillis, (double)kOneDay);
```

Note that `millisInDay` is just getting the "remainder" of that division operation. It would be cleaner if you could do that all in one step (get the quotient and remainder). There are C++ standard library functions for this, such as

https://www.cplusplus.com/reference/cstdlib/div/

I think we have `uprv_div` but unsure.

Reply:
`ClockMath` has a `floorDivide` with a remainder as an out parameter, so we should use that (and that would address the replication issue mentioned by Rich.

Reply:
Note that `localMillis` is a double, which is garbage (timestamps should not be doubles). Do you know if there is a limit on the size of the double?

Note that the max safe integer-valued double, 9007199254740991, is less than the capacity of a uint64_t, 18446744073709551615.